### PR TITLE
Latest notebooks and code

### LINKED ISSUE #7: Add data preparation notebooks to repository
I would like to make sure that all relevant data preparation notebooks are checked in to the repository.

### LINKED ISSUE #12: Final repository cleanup
We should
 - [x] Ensure that the two README.md files are correct and up-to-date
 - [x] Add the latest email notebook
 - [x] Double-check the docker containers to make sure that they are capable of training the models
 - [x] Review the code architecture, readability, &c (James)

### PR DESCRIPTION
**Pull request by jamesmcclain**
Closes #7 
Closes #12

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,88 @@
...
+
+You can overwirte some settings or hyperparameters when you enter the run command to start the training, which will be brefiy introduced in the training section.
+
+Our sample config file can be found at:
+
+### train.spacy and dev.spacy
+
+The `train.spacy` and `dev.spacy` are the data for training and evaluating. The files with `spacy` extension is compilable with training in spaCy and spaCy models. The detialed process of creating the data can be found in our notebooks. In short, the spacy file can be created and read as fellowing.
+
+```
```
Comment (AdeelH) [inline README.md]: ` ``` ` --> ` ```python ` (to add syntax highlighting to the code-block).

File: README.md [hunk 2]
```diff
@@ -1,2 +1,120 @@
...
+3. train.spacy
+4. dev.spacy
+
+### base_config.cfg
+
+`base_config.cfg` is the blueprint of the config files that store all the settings and hyperparameters for training. It is used to generate the real config files, `config.cfg`, by using terminal commands.
+
+We have two types of `base_config.cfg` — one is for light spaCy model, the another one is for spaCy transformer model.
+
+The light cpu `base_config.cfg` can be found at:
```
Comment (lewfish) [inline README.md]: Where can it be found? (You're missing a link here and below)

File: README.md [hunk 3]
```diff
@@ -1,2 +1,120 @@
...
+# Cicero-NLP
+
+
+## The Cicero Database
+
+The [Cicero database](https://cicero.azavea.com/docs/) is a comprehensive, highly-accurate database of elected officials and legislative districts in 9 countries worldwide. It is made from public information that can be found on official government websites, and the information is organizedwell-separated into different categories. Those categories include names, phone numbers, fax numbers, and other contact information plus social media identifiers. The source URLs from official government websites are also incorporated into the dataset. Historically, this large dataset of over 57,000 historical and current officials was collected and managed via  extensive human annotation and data entry efforts. A sample of the Cicero Dataset is shown below.
+
+![ciceroSample](image/README/ciceroSample.png)
+
+## Data Processing
```
Comment (lewfish) [inline README.md:29]: I see you have a separate Docker readme and requirements.txt file in `docker/` but it would be good to link to that from the main README since I am used to seeing setup instructions on a main readme.

File: developing_codes/bio_tagging_address.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+1. insert the address together into the essays in the essay dataset
+2. insert the entites from the CoNLL 2003 entities dataset into the essays in the essay dataset, to increase the difficulty of the task
+
+
+special dependencies:
+1. one essay dataset
+2. CoNLL 2003 entities dataset
+
+'''
+import os
```
Comment (lewfish) [inline developing_codes/bio_tagging_address.py]: In Python, typically the Python builtin dependencies are grouped together, then there's a new line, and then there's a group of 3rd party dependencies like Spacy, and then there is a group of dependencies within your project. eg. https://github.com/azavea/raster-vision/blob/master/rastervision_pytorch_learner/rastervision/pytorch_learner/utils/utils.py#L1-L17

**Review by lewfish — COMMENTED** [review]
Overall, the code looks decent. I didn't review it in any great detail though. Here is some general feedback aside from the inline comments I left.
* I would squash related commits together into a single commit using an interactive rebase which you can read about in the `git` docs. For example, you have several commits about updating the readme. It makes sense to have those separate commits while working on something, but when the PR is ready to review I would squash them into a single "updating readme" commit.
* The directory name `developing_codes` is a bit non-standard. Maybe `src` or `scripts` would be better?
* You have a bunch of data files stored in the repo. That's typically an anti-pattern, but it's ok if it's a really tiny file. I would consider transferring the data to an S3 bucket and then linking to it from the repo. Or if the data is already publicly available, just have your scripts download it locally without checking it into the repo.

File: README.md (hunk 2 above)
Comment (jamesmcclain) [reply README.md]: Agreed, I think that should be made explicit somewhere here.  For reference, I believe that these files can be generated by using the "quickstart" tool found [here](https://spacy.io/usage/training#quickstart).

File: developing_codes/bio_tagging_address.py (hunk 4 above)
Comment (jamesmcclain) [reply developing_codes/bio_tagging_address.py]: Using yapf (for example, `yapf -i bio_tagging_address.py`) should take care of a good percentage of style concerns.

File: developing_codes/bio_tagging_mergeA.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+import spacy
+from sklearn.model_selection import train_test_split
+from spacy.tokens import Doc, DocBin
+from spacy.util import filter_spans
+from tqdm import tqdm
+
+
+def get_parser(parser=argparse.ArgumentParser(
+    description=
+    "to BIO tagging on the pure texts for the mergeModelA, which convers information including names, phone numbers and fax numbers. The output file will be in the format of spaCy Doc file."
```
Comment (AdeelH) [inline developing_codes/bio_tagging_mergeA.py]: You can break up long strings into shorter ones so that the formatter line-wraps them nicely. E.g. `"very long string"` --> `"very " "long " "string"`.

File: developing_codes/bio_tagging_mergeA.py [hunk 6]
```diff
@@ -0,0 +1,222 @@
...
+        # for example, the url address is not valid
+        if pure_text is None:
+            continue
+
+        ruler = nlp.add_pipe("span_ruler")
+        bio_tag_pattern_list = []
+
+        # tag the name
+        # tag the full name, case insensitive by using the `LOWER` attribute
+        bio_tag_pattern_list.append({"label": 'NAME', 'pattern': [{'LOWER': information_unit['first_name'].lower()},\
```
Comment (AdeelH) [inline developing_codes/bio_tagging_mergeA.py]: I think the extra backslash here prevented the formatter from formatting this nicely like the lines below.

**Review by AdeelH — APPROVED** [review]
Apart from some styling nitpicks, this looks good to me. 

I have not tried running the code or read it line by line, but overall it looks well-documented. Thank you for that.

Good job!

**Comment by lewfish** [comment]
I forgot to say that it would be good to run your readmes through a spell checker. There should be one available in VS Code.

**Comment by jamesmcclain** [comment]
We should check to see what the respective license are for the essay dataset and CoNLL2003 to make sure that there are no licensing issues preventing them from being included.  If there are no such issues, I would be in favor of keeping them in the repo (although I admit that that is a debatable position).  @lewfish @AdeelH @ravenouse

**Comment by ravenouse** [comment]
I will check the licenses for these two datasets.

File: developing_codes/bio_tagging_mergeA.py (hunk 5 above)
Comment (ravenouse) [reply developing_codes/bio_tagging_mergeA.py]: Thanks for pointing this out. Could you be more specific on this? 
I searched this [conversation](https://stackoverflow.com/questions/29613487/multiple-lines-in-python-argparse-help-display) about `textwrap`.
But I am not sure if this is your point.

Comment (lewfish) [reply developing_codes/bio_tagging_mergeA.py]: Exmaple: https://github.com/azavea/raster-vision/blob/master/rastervision_core/rastervision/core/data/dataset_config.py#L29-L32

**Comment by jamesmcclain** [comment]
> I will check the licenses for these two datasets.

@ravenouse  @AdeelH @lewfish I think that we have concluded that we will remove these two datasets from the PR before merging.

**Comment by jamesmcclain** [comment]
@ravenouse I think that last remaining change is to remove your WandB key.
